### PR TITLE
Who's turn

### PR DESCRIPTION
**Pull request by xtrembaker**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -9,10 +9,18 @@ jobs:
       - image: circleci/node:7.10
     steps:
       - checkout
-      - run: cd client && yarn install && yarn test
+      - run: cd client && ./bin/ci_running.sh && yarn lint
+      - run: cd client && ./bin/ci_running.sh && yarn test
+  functions_test:
+    docker:
+      - image: circleci/node:7.10
```
Comment (brissaur) [inline .circleci/config.yml]: not compatible with js version of functions right ?

File: client/src/player/sagas.js [hunk 2]
```diff
@@ -43,28 +44,35 @@ function* connectPlayer() {
...
-    result = yield db.collection('players').add({
-        firstname: 'robot3',
-    });
-    playersId.push(result.id);
+/**
+ * @todo Remove that function once multiplayer feature is available
+ * @return {array} of player's
+ */
+export function* createFakePlayers() {
+    const players = [];
```
Comment (brissaur) [inline client/src/player/sagas.js:69]: FYI
return new Array(4).map((elem,index) => yield createFakePlayer(`robot${i + 1}`))
(just for the syntaxe JS // Functional Programming -like, but it wouldn't work here due to the `yield`

File: client/src/table/sagas.js [hunk 3]
```diff
@@ -70,17 +73,27 @@ export function* watchUpdateOnDocumentTable() {
...
+        isFakePlayer: false,
+        pos: 0,
+    });
+    // Create fake player
+    const fakePlayers = yield call(createFakePlayers);
 
-    const players = playersId.map(playerId => ({
-        id: playerId,
+    // for each player, add a position and add them into the list
+    fakePlayers.forEach((player, pos) => players.push({
```
Comment (brissaur) [inline client/src/table/sagas.js]: this is EXACTLY a map :)

File: client/src/table/sagas.js [hunk 4]
```diff
@@ -70,17 +73,27 @@ export function* watchUpdateOnDocumentTable() {
...
+        ...player,
+        pos: pos + 1,
     }));
 
     const document = yield db.collection(COLLECTION_NAME).add({
         players,
         trick: [],
+        state: {},
```
Comment (brissaur) [inline client/src/table/sagas.js]: ah!

File: functions/src/index.js [hunk 5]
```diff
@@ -1,29 +1,65 @@
 import shuffle from 'lodash/shuffle';
-const functions = require('firebase-functions');
+import sortBy from 'lodash/sortBy';
+import { firestore } from 'firebase-functions';
+
+const assignCardsToPlayer = (cards, player) => ({ ...player, cards });
```
Comment (brissaur) [inline functions/src/index.js:5]: COOL :)

File: functions/src/index.js [hunk 6]
```diff
@@ -1,29 +1,65 @@
...
-            players[playerNumber]
+            players[playerNumber],
         );
     }
 
-    return event.data.ref.update({players: players});
-})
+    return event.data.ref.update({
+        players,
+        state: {
```
Comment (brissaur) [inline functions/src/index.js]: ah!

File: functions/src/index.spec.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
+describe('test functions', () => {
```
Comment (brissaur) [inline functions/src/index.spec.js:1]: what for humm... ?
:)

File: functions/src/index.js [hunk 8]
```diff
@@ -1,29 +1,65 @@
 import shuffle from 'lodash/shuffle';
-const functions = require('firebase-functions');
+import sortBy from 'lodash/sortBy';
+import { firestore } from 'firebase-functions';
+
+const assignCardsToPlayer = (cards, player) => ({ ...player, cards });
+
+const documentAllTable = firestore.document('tables/{tableId}');
```
Comment (brissaur) [inline functions/src/index.js]: i don't understand this name it confuses me more than ...

File: functions/src/index.js [hunk 9]
```diff
@@ -1,29 +1,65 @@
...
+    const previousPlayers = sortBy(previousData.players, ['pos']);
+
+    // Trick has changed, we need to select the next player
+    if (currentTrick.length > previousTrick.length) {
+        const previousPlayerId = previousData.general.currentPlayerId;
+        // eslint-disable-next-line max-len
+        const previousPlayerIdx = previousPlayers.findIndex(player => previousPlayerId === player.id);
+        // if previousPlayer was the last one, take the first one
+        const lastPlayerIdx = previousData.players.length - 1;
+        currentPlayerId = (previousPlayerIdx === lastPlayerIdx)
```
Comment (brissaur) [inline functions/src/index.js]: this is a modulo (%)

File: functions/src/index.js [hunk 10]
```diff
@@ -1,29 +1,65 @@
...
+    let currentPlayerId = null;
+    const previousPlayers = sortBy(previousData.players, ['pos']);
+
+    // Trick has changed, we need to select the next player
+    if (currentTrick.length > previousTrick.length) {
+        const previousPlayerId = previousData.general.currentPlayerId;
+        // eslint-disable-next-line max-len
+        const previousPlayerIdx = previousPlayers.findIndex(player => previousPlayerId === player.id);
+        // if previousPlayer was the last one, take the first one
+        const lastPlayerIdx = previousData.players.length - 1;
```
Comment (brissaur) [inline functions/src/index.js]: should we hardcode 4 as number of players?

File: .circleci/config.yml (hunk 1 above)
Comment (xtrembaker) [reply .circleci/config.yml]: Done

File: client/src/player/sagas.js (hunk 2 above)
Comment (xtrembaker) [reply client/src/player/sagas.js:69]: Ok thanks for the tips !

File: functions/src/index.js (hunk 5 above)
Comment (xtrembaker) [reply functions/src/index.js:5]: Top :)

File: client/src/table/sagas.js (hunk 3 above)
Comment (xtrembaker) [reply client/src/table/sagas.js]: Done

File: client/src/table/sagas.js (hunk 4 above)
Comment (xtrembaker) [reply client/src/table/sagas.js]: Done

File: functions/src/index.js (hunk 8 above)
Comment (xtrembaker) [reply functions/src/index.js]: Fair enough

File: functions/src/index.js (hunk 6 above)
Comment (xtrembaker) [reply functions/src/index.js]: Done

File: functions/src/index.js (hunk 10 above)
Comment (xtrembaker) [reply functions/src/index.js]: Done

File: functions/src/index.js (hunk 9 above)
Comment (xtrembaker) [reply functions/src/index.js]: Done

File: functions/src/index.spec.js (hunk 7 above)
Comment (xtrembaker) [reply functions/src/index.spec.js:1]: for test !! :)

**Pull request merged**